lib/babe/build.go

```diff
@@ -246,3 +248,63 @@ func extrinsicsToBody(txs []*transaction.ValidTransaction) (*types.Body, error)
 
 	return types.NewBodyFromExtrinsics(extrinsics)
 }
+
+func determineError(res []byte) string {
+	log.Error("[babe] build block apply extrinsic", "error", res)
+	var errTxt strings.Builder
+
+	// when res[0] == 0x01 it is an apply error
+	if res[0] == 1 {
+		errTxt.WriteString("Apply error, type: ")
+		if bytes.Equal(res[1:], []byte{0}) {
+			errTxt.WriteString("NoPermission")
+		}
+		if bytes.Equal(res[1:], []byte{1}) {
+			errTxt.WriteString("BadState")
+		}
+		if bytes.Equal(res[1:], []byte{2}) {
+			errTxt.WriteString("Validity")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 0}) {
+			errTxt.WriteString("Call")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 1}) {
+			errTxt.WriteString("Payment")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 2}) {
+			errTxt.WriteString("Future")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 3}) {
+			errTxt.WriteString("Stale")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 4}) {
+			errTxt.WriteString("BadProof")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 5}) {
+			errTxt.WriteString("AncientBirthBlock")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 6}) {
+			errTxt.WriteString("ExhaustsResources")
+		}
+		if bytes.Equal(res[1:], []byte{2, 0, 7}) {
+			errTxt.WriteString("Custom")
+		}
+		if bytes.Equal(res[1:], []byte{2, 1, 0}) {
+			errTxt.WriteString("CannotLookup")
+		}
+		if bytes.Equal(res[1:], []byte{2, 1, 1}) {
+			errTxt.WriteString("NoUnsignedValidator")
+		}
+		if bytes.Equal(res[1:], []byte{2, 1, 2}) {
+			errTxt.WriteString("Custom")
+		}
```

Comment:
I wanted to do all this in a switch, but I was getting `invalid case []byte literal in switch` errors, so I'm using if statements, is this too ugly?  Any suggestions?

Reply:
I don't mind it, if a switch isn't working there isn't much you can do

Reply:
same as above, this works for me as is but if you see a potential optimization, wouldn't hurt to open an issue